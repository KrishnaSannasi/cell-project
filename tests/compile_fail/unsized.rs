use cell_project::cell_project as project;
use std::cell::Cell;

pub struct Foo {
    bar: i32,
    quax: [u8],
}

fn check_unsized_field(foo: &Foo) -> &[u8] {
    &foo.quax
}

fn project(cell: &Cell<Foo>) -> &Cell<[u8]> {
    project!(Foo, cell.quax)
}

fn main() {}
