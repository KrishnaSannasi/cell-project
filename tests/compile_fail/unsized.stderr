error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
  --> $DIR/unsized.rs:14:5
   |
14 |     project!(Foo, cell.quax)
   |     ^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
   | 
  ::: $WORKSPACE/src/lib.rs:9:33
   |
9  | pub unsafe fn project_unchecked<T, F>(cell: &Cell<T>, field: *const F, offset: usize) -> &Cell<F> {
   |                                 - required by this bound in `cell_project::project_unchecked`
   |
   = help: within `Foo`, the trait `std::marker::Sized` is not implemented for `[u8]`
   = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
   = note: required because it appears within the type `Foo`
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
  --> $DIR/unsized.rs:14:5
   |
14 |     project!(Foo, cell.quax)
   |     ^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
   | 
  ::: $WORKSPACE/src/lib.rs:9:36
   |
9  | pub unsafe fn project_unchecked<T, F>(cell: &Cell<T>, field: *const F, offset: usize) -> &Cell<F> {
   |                                    - required by this bound in `cell_project::project_unchecked`
   |
   = help: the trait `std::marker::Sized` is not implemented for `[u8]`
   = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
